kernel/hConstr.ml

```diff
@@ -0,0 +1,370 @@
+   instead use physical equality and make fresh int refs
+
+   XXX this is terrible
+   instead generate a unique id for each binder crossed
+   and have a field in hconstr containing the id for each rel
+   the field participates to both hash and equality
+   the rel is still there so liftings are still handled
+
+   or even generate unique id only for the case binders where we don't have the type
+*)
+
+module Self = struct
+
+type t = {
+  self : constr;
+  kind : (t,t,Sorts.t,UVars.Instance.t,Sorts.relevance) kind_of_term;
+  isRel : int (* 0 -> not a rel, otherwise unique identifier of that binder *);
+  hash : int;
+  mutable refcount : int;
+}
+
+let equal a b =
+  a.isRel == b.isRel
+  && hasheq_kind a.kind b.kind
+
+let hash x = x.hash
+
+end
+
+include Self
+
+let self x = x.self
+
+let refcount x = x.refcount
+
+module Tbl = struct
+  type key = t
+
+  (* We expose the API of Hashtbl but implemented using Int.Map ref.
+     For unclear reasons this is dramatically faster on an artificial example
+     (balanced binary tree whose leaves are all different primitive ints,
+      such that there is no sharing).
+     It is a bit slower in the real world.
+     It seems that hashtbl produces overly large buckets which then need to be linearly scanned.
+     hconsing doesn't seem to have this problem,
+     perhaps because of differences between hashtbl and our hashset implementation. *)
+  type 'a t = (key * 'a) list Int.Map.t ref
+
+  let create _ = ref Int.Map.empty
+
+  let add tbl key v =
+    tbl := Int.Map.update key.hash (function
+        | None -> Some [(key,v)]
+        | Some l -> Some ((key,v)::l))
+        !tbl
+
+  let find_opt tbl key =
+    match Int.Map.find_opt key.hash !tbl with
+    | None -> None
+    | Some l ->
+      List.find_map (fun (k',v) ->
+          if equal key k' then Some v else None)
+        l
+
+  type stats = {
+    hashes : int;
+    bindings : int;
+    most_collisions : int;
+  }
+
+  let empty_stats = {
+    hashes = 0;
+    bindings = 0;
+    most_collisions = 0;
+  }
+
+  let stats tbl =
+    Int.Map.fold (fun _ l acc ->
+        let len = List.length l in
+        {
+          hashes = acc.hashes + 1;
+          bindings = acc.bindings + len;
+          most_collisions = max acc.most_collisions len;
+        }
+      )
+      !tbl
+      empty_stats
+
+end
+
+type local_env = {
+  (* unique identifiers for each binder crossed *)
+  rels : int Range.t;
+  (* global counter *)
+  cnt : int ref;
+  assum_uids : int Tbl.t;
+  (* the surrounding table is for the body, the inner table for the type *)
+  letin_uids : int Tbl.t Tbl.t;
+}
+
+let empty_env () = {
+  rels = Range.empty;
+  cnt = ref 0;
+  assum_uids = Tbl.create 47;
+  letin_uids = Tbl.create 47;
+}
+
+let push_unknown_rel env =
+  incr env.cnt;
+  { env with rels = Range.cons !(env.cnt) env.rels }
+
+let push_assum t env =
+  let uid = match Tbl.find_opt env.assum_uids t with
+  | Some uid -> uid
+  | None ->
+    incr env.cnt;
```

Comment:
You should abstract the uid generation in a standalone function and add there the assertion that the uid never wraps around, rather than in the of_constr function.